lib/srv/desktop/rdp/rdpclient/src/rdpdr/scard.rs

```diff
@@ -1853,3 +2192,577 @@ fn debug_print_payload(payload: &mut Payload) {
+                148, 106, 24, 54, 54, 15, 193, 104, 3, 193, 248, 118, 255, 103, 245, 208, 202, 91,
+                110, 91, 229, 246, 173, 240, 111, 25, 182, 9, 180, 245, 147, 241, 247, 141, 222, 5,
+                46, 146, 194, 184, 7, 254, 106, 167, 126, 27, 233, 33, 7, 112, 54, 209, 9, 195,
+                198, 17, 208, 79, 57, 163, 61, 128, 82, 212, 65, 5, 119, 221, 202, 75, 227, 70, 77,
+                2, 197, 239, 8, 29, 71, 101, 2, 129, 129, 0, 141, 75, 99, 52, 181, 42, 232, 22, 82,
+                120, 119, 201, 255, 122, 220, 146, 225, 193, 162, 102, 82, 30, 94, 140, 197, 50,
+                59, 122, 2, 92, 75, 64, 178, 230, 209, 216, 171, 40, 172, 143, 128, 77, 160, 241,
+                130, 40, 205, 123, 241, 181, 38, 13, 24, 215, 218, 53, 217, 82, 125, 201, 153, 141,
+                149, 236, 191, 149, 137, 125, 208, 56, 69, 217, 228, 65, 85, 148, 234, 30, 115, 31,
+                81, 234, 98, 250, 222, 165, 236, 164, 56, 19, 34, 29, 150, 172, 118, 228, 179, 91,
+                26, 208, 186, 161, 49, 218, 225, 211, 204, 48, 207, 193, 226, 158, 174, 105, 177,
+                227, 28, 132, 109, 252, 218, 102, 20, 175, 152, 91, 201, 168,
+            ],
+            "68971585".to_string(),
+        )
+    }
+
+    fn to_payload(e: &dyn Encode) -> Payload {
+        Payload::new(e.encode().unwrap())
+    }
+
+    fn test_ioctl(
+        established_ctxs: u32,
+        connect_scard_to_ctx: Option<u32>,
+        ctl_code: IoctlCode,
+        payload: &dyn Encode,
+        expected: Vec<u8>,
+    ) {
+        let mut c = client();
+
+        for _ in 0..established_ctxs {
+            c.contexts.establish();
+        }
+        if let Some(connect_scard_to_ctx) = connect_scard_to_ctx {
+            connect_scard(&mut c, connect_scard_to_ctx);
+        }
+
+        let (code, res) = c.ioctl(ctl_code, &mut to_payload(payload)).unwrap();
+        assert_eq!(0, code);
+        assert_eq!(expected, res);
+    }
+
+    /// Connects a piv::Card to the client's internal context cache
+    /// (on the context corresponding to context_value). This is a manual
+    /// way of doing what test_scard_ioctl_connectw does to the Client's
+    /// internal state.
+    fn connect_scard(c: &mut Client, context_value: u32) {
+        let ctx = c.contexts.get(context_value).unwrap();
+        ctx.connect(
+            Context {
+                length: 4,
+                value: context_value,
+            },
+            c.uuid,
+            &c.cert_der,
+            &c.key_der,
+            c.pin.clone(),
+        )
+        .unwrap();
+    }
+
+    #[test]
+    fn test_accessstartedevent() {
+        test_ioctl(
+            0,
+            None,
+            IoctlCode::SCARD_IOCTL_ACCESSSTARTEDEVENT,
+            &ScardAccessStartedEvent_Call {
+                _unused: 3234823568,
+            },
+            vec![
+                1, 16, 8, 0, 204, 204, 204, 204, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
```

Comment:
I guess that the raw values here are correct but I was wondering how did you come up with these values? Are those from the specs or from your own tests? I'm just asking as if I ever want to change anything here it might be great to have it written down how to get these values 😄

Reply:
These are just raw vectors that I copied from the debug log output. They get removed and replaced with more human-comprehensible structures in a follow up PR:

https://github.com/gravitational/teleport/pull/16862/files#diff-369d1be40058565e2f51b7ba105d18945506534aba344edd336f0fbe8585ff75